BUG: Let 0-D arrays of Python timedelta convert to np.timedelta64.

tylerjereddy opened the pull request:
Fixes #11151.

It is now possible to convert 0 dimensional `datetime.timedelta` NumPy object arrays to `np.timedelta64` 0 dimensional NumPy arrays. 

The linked issue does not specify other 0 dimensional NumPy object array-related datetime issues so this patch initially aims only to fix that specific reported scenario, but I've added a comment in the C source suggesting that other 0 dimensional object array type checks / handling might be needed in the future.

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3746,6 +3746,27 @@ recursive_find_object_datetime64_type(PyObject *obj,
     }
 }
 
+/* handler function for PyDelta values
```
We generally prefer the first line of these to be blank,
```
/*
 * blah
 */
```

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3746,6 +3746,27 @@ recursive_find_object_datetime64_type(PyObject *obj,
+ */
+static int
+delta_checker(PyObject *obj,
+	      PyArray_DatetimeMetaData *meta)
```
the function header fits nicely on a single line.

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3746,6 +3746,27 @@ recursive_find_object_datetime64_type(PyObject *obj,
+    tmp_meta.num = 1;
+
+    /* Combine it with 'meta' */
+    if (compute_datetime_metadata_greatest_common_divisor(meta,
```
might as well move all the arguments to the next line
```
if (compute_datetime_metadata_greatest_common_divisor(
        meta, &tmp_meta, meta, 0, 0) < 0) {
```

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3783,6 +3804,29 @@ recursive_find_object_timedelta64_type(PyObject *obj,
+            Py_ssize_t ndim;
+            ndim = PyArray_NDIM((PyArrayObject *)obj);
+            if (ndim == 0) {
+                /* special handling of 0 dimensional
```
See above.

charris replied:
might also justify a bit to lessen the number of lines.

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3783,6 +3804,29 @@ recursive_find_object_timedelta64_type(PyObject *obj,
+                 * object using [()], but not by using
+                 * __getitem__(integer) approaches
+                 */
+                PyObject *item, *meth, *args;
```
Blank line after declarations.

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3783,6 +3804,29 @@ recursive_find_object_timedelta64_type(PyObject *obj,
+                meth = PyObject_GetAttrString(obj, "__getitem__");
+                args = Py_BuildValue("(())");
+                item = PyObject_CallObject(meth, args);
+                /* NOTE: may need other type checks
```
See above.

charris commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3783,6 +3804,29 @@ recursive_find_object_timedelta64_type(PyObject *obj,
         }
+        else {
+            Py_ssize_t ndim;
+            ndim = PyArray_NDIM((PyArrayObject *)obj);
```
Blank line after declarations.
The return type of `PyArray_NDIM` is `int`, any reason to change that?

charris replied:
I note that up above `arr` is already the cast of `obj`, so why not
```
            if (PyArray_NDIM(arr) == 0) {
```

eric-wieser commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -3746,6 +3746,27 @@ recursive_find_object_datetime64_type(PyObject *obj,
+ * NumPy array
+ */
+static int
+delta_checker(PyObject *obj, PyArray_DatetimeMetaData *meta)
```
Why does this function not touch `obj`?

tylerjereddy replied:
The function is nothing more than a container for the code I abstracted from the original function because it is now duplicated.

I think a lot of the datetime handling happens at a metadata level & can indeed just remove the `obj` argument here I think.

tylerjereddy committed:
BUG: 0D object array to timedelta

* it is now possible to convert
a 0 dimensional object array to
a 0 dimensional NumPy timedelta
array

* the single datetime.timedelta
entry in the 0 D NumPy array
is accessed using __getitem__(())
because integer indexing doesn't
work on 0D NumPy arrays

charris commented:
Thanks Tyler.
